File: src/main/resources/js/ticTacToe.js
var ws;
var player;
function init() {
    ws = new WebSocket("ws://localhost:9001/");
    ws.onopen = function (evt) {
        publishEvent(evt, "opened");
        ws.send("join");
    };
    ws.onmessage = ordinaryOnMessage;
    ws.onclose = function (evt) {
        publishEvent(evt, "closed")
    };
    function bindClick() {
        var trs = document.getElementsByTagName("tr");
        for (i = 0; i < trs.length; i++) {
            var tds = trs[i].children;
            for (j = 0; j < tds.length; j++) {
                tds[j].addEventListener("click", makeTurn(j, i));
            }
        }
    }
    bindClick();
}
function ordinaryOnMessage(evt) {
    publishEvent(evt, "message");
    json = JSON.parse(evt.data);
    if (json.field) {
        drawField(json.field)
    }
    if (json.ch) {
        player = json.ch;
    }
}
function waitingStatusOnMessage (elem) {
    return function(evt) {
        publishEvent(evt, "message");
        json = JSON.parse(evt.data);
        if (json.status == 'success' || json.status == 'Win' || json.status == 'Tie') {
            visualize(elem, player);
        }
        ws.onmessage = ordinaryOnMessage;
    }
}
function visualize(elem, sign, s) {
    if(sign == '-') {
        return false;
    }
    while(elem.firstChild) {
        elem.removeChild(elem.firstChild);
    }
    var size;
    if(s) {
        size = s;
    } else {
        size = 90;
    }
    var pic = new Image(size,size);
    if(sign == 'x') {
        pic.src  = 'images/x.png';
        elem.appendChild(pic);
    }
    if(sign == 'o') {
        pic.src  = 'images/o.png';
        elem.appendChild(pic);
    }
}
function drawField(field) {
    var trs = document.getElementsByTagName("tr");
    for (i = 0; i < trs.length; i++) {
        var tds = trs[i].children;
        for (j = 0; j < tds.length; j++) {
            visualize(tds[j],field[i * 3 + j]);
        }
    }
}
function makeTurn(x_move, y_move) {
    return function aux() {
        ws.onmessage = waitingStatusOnMessage(this);
        ws.send(JSON.stringify({x: x_move, y: y_move}));
    }
}
function publishEvent(evt, type) {
    publish(evt.data ? evt.data : "", type);
}
function publish(message, type) {
    var text;
    if(type == "opened") {
        text = "Wait for your opponent";
    }
    if(type == "message") {
        var mess = JSON.parse(message);
        if(mess.status == "success") {
            text = "Wait for your turn";
        }
        if(mess.status == "failure") {
            text = mess.message;
        }
        if(mess.status == "New") {
            text = "Game began, your turn";
        }
        if(mess.status == "Game") {
            text = "Your turn";
        }
        if(mess.status == "Win") {
            text = "Congratulations, you win!";
        }
        if(mess.status == "Lose") {
            text = "Bad luck, you lose!";
        }
        if(mess.status == "Tie") {
            text = "Game over, it is tie";
        }
        if(mess.ch) {
            sign = document.getElementById("sign");
            visualize(sign, mess.ch, 40);
        }
    }
    doc = document.getElementById("messages");
    doc.innerHTML = text;
}
window.addEventListener("load", init, false);

File: app/js/src/main/scala/TicTacToe.scala
import org.scalajs.dom._
import shared._

import scala.scalajs.js.annotation.JSExport

@JSExport
object TicTacToe {

  @JSExport
  def main(body: html.Div) = {
    var player = '-'
    val ws = new WebSocket("ws://localhost:9001")

    ws.onopen = (e: Event) => {
      publishEvent(Opened)
      ws.send(upickle.write(Start))
    }

    ws.onmessage = ordinaryOnMessage _

    ws.onclose = (evt: Event) => publishEvent(Closed)

    def makeTurn(td: Node, x: Int, y: Int) = {
      ws.onmessage = waitingStatusOnMessage(td)
      ws.send(upickle.write(GameMove(x,y)))
    }

    def ordinaryOnMessage(evt: MessageEvent) = {
      val ex = upickle.read[SharedExchange](evt.data.toString)
      publishEvent(ex)
      ex match {
        case GameState(field, _) => drawField(field)
        case Player(ch) =>
          player = ch
          Common.visualize(document.getElementById("sign"), player, 30)
        case _ => ""
      }
    }

    def waitingStatusOnMessage (elem:Node) = (evt: MessageEvent) => {
      val ex = upickle.read[SharedExchange](evt.data.toString)
      publishEvent(ex)
      ex match {
        case Success(_) | GameState(_,Win) | GameState(_,Tie) => Common.visualize(elem, player, 90)
        case _ => ""
      }
      ws.onmessage = ordinaryOnMessage _
    }

    def bindClick() = Common.transformField(
      (td: Node, i: Int, j: Int) => td.addEventListener[MouseEvent]("click", (e: MouseEvent) => makeTurn(td, j, i))
    )

    def drawField(field: String) = Common.transformField (
      (td: Node, i: Int, j: Int) => Common.visualize(td, field(3*i+j), 90)
    )

    def publishEvent(message: SharedExchange) =
      document.getElementById("messages").innerHTML = message match {
        case Success(_) => "Wait for your turn"
        case Failure(m, _) => m
        case GameState(_, status) => status match {
          case New => "Game began, your turn"
          case Game => "Your turn"
          case Win => "Congratulations, you win!"
          case Lose => "Bad luck, you lose!"
          case Tie => "Game over, it is tie"
          case WrongMove => "Wrong Move, bastard"
        }

        case Opened => "Wait for your turn"
        case Closed => "Closed, reload page please"

        case _ => ""
      }

    body.appendChild(Common.createField(3))
    bindClick()
  }

}


